Casting object pointers/void pointers to function pointers not allowed in standard C (#1)
jeremy-rifkin/c-cpp-trivia: README.md

- C and C++ support a set of
  [digraph and trigraph](https://en.wikipedia.org/wiki/Digraphs_and_trigraphs#C) tokens to
  accommodate certain archaic keyboards. Trigraphs were removed from C++ in C++17.

> - ISO C forbids conversion of a function pointer to an object pointer type:

Reviewer: Would be good to indicate the conversion is forbidden in both directions
```suggestion
- ISO C forbids conversion between a function and object pointers:
```
I quickly looked for a standard reference for this but I couldn't find it. If you have one it'd be good to include, no worries if not!

Author: It's in [the standard at 6.3.2.3p8](http://port70.net/~nsz/c/c99/n1256.html#6.3.2.3p8)

> [8](http://port70.net/~nsz/c/c99/n1256.html#6.3.2.3p8) A pointer to a function of one type may be converted to a pointer to a function of another type and back again; the result shall compare equal to the original pointer. If a converted pointer is used to call a function whose type is not compatible with the pointed-to type, the behavior is undefined.

In this section it only states what IS allowed but does not explicitly disallow it, so this is weird but true.

  ```cpp

>   void (*func_ptr)() = dlsym(mylib, "func"); // gcc yields a warning with standard C17 in pedantic mode

Reviewer: Is this a new requirement in C17? I see this warning appearing for other C versions as well. If not new in C17, would be good to not overspecify.

Author: You are right, with GCC this is actually a requirement for any standard but only when you use "-pedantic"

  ```

>   However, if taking the address to the function pointer first, then casting to `void**` and finally dereferencing this pointer again, makes it work without warnings:

Reviewer: I think there's room for improvement on this sentence, maybe it could be reworded something like this?
```suggestion
  Casting the pointer from `dlsym` is forbidden but a trick to do this without warnings is to do the conversion on the left hand side as `&func_ptr` is an object pointer:
```
